### PR TITLE
Fix close deadlock and Sub type error

### LINKED ISSUE #10: Already created subscriptions should be killed when subscription creation process fails
https://github.com/libp2p/go-eventbus/blob/8acb6225a80c569b998b693d435c5cbe86e01eab/basic.go#L155

### PR DESCRIPTION
**Pull request by magik6k**
Fixes libp2p/go-eventbus#10 and libp2p/go-libp2p#1704

Bench:
```
benchmark                       old ns/op     new ns/op     delta
BenchmarkSubs-32                498649        501377        +0.55%
BenchmarkEmits-32               962425        919298        -4.48%
BenchmarkMsgs-32                933676        990440        +6.08%
BenchmarkOneToMany-32           71033         70820         -0.30%
BenchmarkManyToOne-32           130073        131921        +1.42%
BenchmarkMs1e2m4-32             2078          2228          +7.22%
BenchmarkMs1e0m6-32             387           427           +10.34%
BenchmarkMs0e0m6-32             187           187           +0.00%
BenchmarkStatefulMs1e0m6-32     405           353           -12.84%
BenchmarkStatefulMs0e0m6-32     202           201           -0.50%
BenchmarkMs0e6m0-32             5355          4964          -7.30%
BenchmarkMs6e0m0-32             4241          6817          +60.74%

benchmark                       old allocs     new allocs     delta
BenchmarkSubs-32                8              10             +25.00%
BenchmarkEmits-32               7              7              +0.00%
BenchmarkMsgs-32                0              0              +0.00%
BenchmarkOneToMany-32           8              10             +25.00%
BenchmarkManyToOne-32           6              6              +0.00%
BenchmarkMs1e2m4-32             0              0              +0.00%
BenchmarkMs1e0m6-32             0              0              +0.00%
BenchmarkMs0e0m6-32             0              0              +0.00%
BenchmarkStatefulMs1e0m6-32     0              0              +0.00%
BenchmarkStatefulMs0e0m6-32     0              0              +0.00%
BenchmarkMs0e6m0-32             7              7              +0.00%
BenchmarkMs6e0m0-32             9              13             +44.44%

benchmark                       old bytes     new bytes     delta
BenchmarkSubs-32                355           520           +46.48%
BenchmarkEmits-32               254           256           +0.79%
BenchmarkMsgs-32                21            27            +28.57%
BenchmarkOneToMany-32           370           566           +52.97%
BenchmarkManyToOne-32           169           129           -23.67%
BenchmarkMs1e2m4-32             0             0             +0.00%
BenchmarkMs1e0m6-32             0             0             +0.00%
BenchmarkMs0e0m6-32             0             0             +0.00%
BenchmarkStatefulMs1e0m6-32     0             0             +0.00%
BenchmarkStatefulMs0e0m6-32     0             0             +0.00%
BenchmarkMs0e6m0-32             575           576           +0.17%
BenchmarkMs6e0m0-32             429           1142          +166.20%
```

### CODE REVIEW

File: basic.go [hunk 1]
```diff
@@ -104,22 +104,38 @@ func (s *sub) Out() <-chan interface{} {
 }
 
 func (s *sub) Close() error {
-	close(s.ch)
+	stop := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case <-s.ch:
```
Comment (raulk) [inline basic.go:111]: This might be racy. If there are many items queued in s.ch, the code below may end up closing stop before we fully drain the channel.

Comment (raulk) [reply basic.go:111]: How about we block Close until the channel is closed and empty? We can drain the channel concurrently in a goroutine (like here) while we remove it from nodes, but we block the return until the channel is empty and closed. Makes for more predictable behaviour.

We should also add a subscriber option so the consumer can decide if they want pending items drained or not on closure (this would enable certain test scenarios with for range ch).

Comment (magik6k) [reply basic.go:111]: Removing stuff from nodes happens under write lock, which we can only take if the channel is drained.

We don't really care to drain the channel fully (if buffered), we only need to make sure we remove it from relevant nodes, and let GC take care of the rest

**Comment by Kubuxu** [comment]
Would be great to have a test for libp2p/go-eventbus#10.

**Comment by magik6k** [comment]
(done)

**Comment by Kubuxu** [comment]
I've improved the test for libp2p/go-eventbus#10

File: basic.go (hunk 1 above)
Comment (Kubuxu) [reply basic.go:111]: IMO let's merge it, it is better than it was. If there is an edge case we will discover it.

**Review by Kubuxu — APPROVED** [review]

File: basic.go (hunk 1 above)
Comment (Stebalien) [reply basic.go:111]: I'd also rather fully drain the channel, mostly because it entirely removes the "stop" channel, but either way works. let's get this fixed.
